### DISCUSSION TITLE
`EpiNow2::discretise` is masking `epiparameter::discretise`

### DISCUSSION THREAD

**Discussion by avallecam** [epiforecasts/EpiNow2#582] [General]
reporting that new function `discretise()` in https://github.com/epiforecasts/EpiNow2/pull/504 is masking {epiparameter} one. tagging @jamesmbaazam in case informative for interoperability purposes.

``` r
library(epiparameter)
library(EpiNow2)
#> 
#> Attaching package: 'EpiNow2'
#> The following object is masked from 'package:epiparameter':
#> 
#>     discretise
#> The following object is masked from 'package:stats':
#> 
#>     Gamma

ebola_serial <- epidist_db(
  disease = "ebola",
  epi_dist = "serial",
  single_epidist = TRUE
)
#> Using WHO Ebola Response Team, Agua-Agum J, Ariyarajah A, Aylward B, Blake I,
#> Brennan R, Cori A, Donnelly C, Dorigatti I, Dye C, Eckmanns T, Ferguson
#> N, Formenty P, Fraser C, Garcia E, Garske T, Hinsley W, Holmes D,
#> Hugonnet S, Iyengar S, Jombart T, Krishnan R, Meijers S, Mills H,
#> Mohamed Y, Nedjati-Gilani G, Newton E, Nouvellet P, Pelletier L,
#> Perkins D, Riley S, Sagrado M, Schnitzler J, Schumacher D, Shah A, Van
#> Kerkhove M, Varsaneux O, Kannangarage N (2015). "West African Ebola
#> Epidemic after One Year — Slowing but Not Yet under Control." _The New
#> England Journal of Medicine_. doi:10.1056/NEJMc1414992
#> <https://doi.org/10.1056/NEJMc1414992>.. 
#> To retrieve the short citation use the 'get_citation' function

# expected error
discretise(ebola_serial)
#> Error in discretise(ebola_serial): Can only discretise a <dist_spec>.

# epiparameter
epiparameter::discretise(ebola_serial)
#> Disease: Ebola Virus Disease
#> Pathogen: Ebola Virus
#> Epi Distribution: serial interval
#> Study: WHO Ebola Response Team, Agua-Agum J, Ariyarajah A, Aylward B, Blake I,
#> Brennan R, Cori A, Donnelly C, Dorigatti I, Dye C, Eckmanns T, Ferguson
#> N, Formenty P, Fraser C, Garcia E, Garske T, Hinsley W, Holmes D,
#> Hugonnet S, Iyengar S, Jombart T, Krishnan R, Meijers S, Mills H,
#> Mohamed Y, Nedjati-Gilani G, Newton E, Nouvellet P, Pelletier L,
#> Perkins D, Riley S, Sagrado M, Schnitzler J, Schumacher D, Shah A, Van
#> Kerkhove M, Varsaneux O, Kannangarage N (2015). "West African Ebola
#> Epidemic after One Year — Slowing but Not Yet under Control." _The New
#> England Journal of Medicine_. doi:10.1056/NEJMc1414992
#> <https://doi.org/10.1056/NEJMc1414992>.
#> Distribution: discrete gamma
#> Parameters:
#>   shape: 2.188
#>   scale: 6.490
```

<sup>Created on 2024-02-26 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

EpiNow2 Version:
`remotes::install_github("epiforecasts/EpiNow2@dist-interface")`

**Comment by jamesmbaazam** [comment]
Thanks, Andree. Worth noting.

In the past, I've dealt with such NAMESPACE conflicts explicitly with the [conflicted package](https://conflicted.r-lib.org/).

**Comment by avallecam** [comment]
> In the past, I've dealt with such NAMESPACE conflicts explicitly with the [conflicted package](https://conflicted.r-lib.org/).

yes, explicit NAMESPACE too. However, would it be viable or sensible to rename `EpiNow2::discretise()` in order to avoid a conflict like this? From a user experience, I find this a minor blocker.

**Reply by seabbs** [reply]
Or alternatively rename it in `epiparameter`?

**Reply by avallecam** [reply]
yes, also fair. I requested this here given the living PR and most recent implementation.

**Reply by jamesmbaazam** [reply]
@avallecam Why is explicitly namespacing a blocker? It's idiomatic to explicitly call a function from a package if it is masked by another. Isn't it?

**Reply by avallecam** [reply]
yes, we can do that. actually, [we agreed](https://github.com/epiverse-trace/episoap/issues/91) and try to use `::` for explicit namespace callings among our materials.

From a dev or interoperability point of view, do we want to allow conflicts for new functions coming into an ecosystem? or when possible, avoid function name conflict within an ecosystem? and less permissive to conflicts with {base} and {stats}? `Gamma()` and `discretise()` could be an example of assessing equally or differently for conflicts with functions from {stats} and an interoperable ecosystem.

A reference could be the fourth bullet suggestion in [ropensci dev guide (on function and argument naming)](https://devguide.ropensci.org/pkg_building.html#function-and-argument-naming). (edit: website re-rendered and link needed an update)

**Reply by avallecam** [reply]
if tidyverse may work as a reference, it looks like having a few conflicts with base and stats (see [output in this reply](https://community.rstudio.com/t/functions-with-the-same-identical-names-in-r-packages/43680/5)), but don't know if within, and also I do not think necessary to quantify it for this aims.

**Comment by seabbs** [comment]
Thanks for this. @avallecam 

I've ported to a discussion as this doesn't seem to be at the level of an issue for `EpiNow2` as it stands.

**Reply by seabbs** [reply]
Also note that the `EpiNow2` funcitonality for this is likely to be depreciated in favour of a new dependency in light of the work in https://www.medrxiv.org/content/10.1101/2024.01.12.24301247v1

**Reply by avallecam** [reply]
Thank you for redirecting this to a more appropriate place. Also, good to know the future plans for this feature. I'm ok to keep suggesting specific namespace for this conflict.

**Reply by avallecam** [reply]
Second thoughts on this, possibly even the temporality of a function may not be an appropriate argument to allow a temporal conflict. In any case, a feature like this always opens topics to discuss with users.

**Comment by avallecam** [comment]
After several trials, we did not have reports of function conflicts, so I find it appropriate to close this. Thanks for the guidance and references shared.
